control-plane/csi-controller/src/controller.rs

```diff
@@ -0,0 +1,726 @@
+        }
+
+        match args.volume_capability {
+            Some(c) => {
+                check_volume_capabilities(&[c])?;
+            }
+            None => {
+                return Err(Status::invalid_argument("Missing volume capability"));
+            }
+        };
+
+        let node_id = normalize_hostname(args.node_id);
+        let volume_id = args.volume_id.to_string();
+
+        // Check if the volume is already published.
+        let volume = MayastorApiClient::get_client()
+            .get_volume(&volume_id)
+            .await?;
+        let uri = if let Some(state) = volume.state.as_ref() {
+            let curr_proto = state.protocol.to_string();
+
+            // Volume is aready published, make sure the protocol matches and get URI.
+            if curr_proto != "none" {
+                if curr_proto != *args.volume_context.get("protocol").unwrap().to_string() {
+                    let m = format!(
+                        "Volume {} already shared via different protocol: {:?}",
+                        volume_id, state.protocol,
+                    );
+                    error!("{}", m);
+                    return Err(Status::failed_precondition(m));
+                }
+                let uri = get_volume_share_uri(&volume)?;
+                debug!("Volume {} already published at {}", volume_id, uri);
+                uri
+            } else {
+                // Volume is not published.
+                let v = MayastorApiClient::get_client()
+                    .publish_volume(&volume_id, &node_id, protocol)
+                    .await?;
+                let uri = get_volume_share_uri(&v)?;
+                debug!("Volume {} successfully published at {}", volume_id, uri);
+                uri
+            }
+        } else {
+            let m = format!("Volume {} is missing current state", volume_id);
+            error!("{}", m);
+            return Err(Status::internal(m));
+        };
+
+        // Prepare the context for the Mayastor Node CSI plugin.
+        let mut publish_context = HashMap::new();
+        publish_context.insert("uri".to_string(), uri);
+
+        if let Some(io_timeout) = args.volume_context.get(volume_opts::IO_TIMEOUT) {
+            publish_context.insert(volume_opts::IO_TIMEOUT.to_string(), io_timeout.to_string());
+        }
+
+        debug!(
+            "Publish context for volume {}: {:?}",
+            volume_id, publish_context
+        );
+        Ok(Response::new(ControllerPublishVolumeResponse {
+            publish_context,
+        }))
+    }
+
+    #[instrument]
+    async fn controller_unpublish_volume(
+        &self,
+        request: tonic::Request<ControllerUnpublishVolumeRequest>,
+    ) -> Result<tonic::Response<ControllerUnpublishVolumeResponse>, tonic::Status> {
+        let args = request.into_inner();
+
+        debug!("Request to unpublish volume: {:?}", args);
+        // Check if node exists.
```

Comment:
We don't seem to be checking if the node exists. Should we be doing so?

Reply:
Yes, I might have overlooked this check - we also need to assure that node is published on the same node as in the request, so we need to add GET node info here.